README
Cave::Wrapper

"cave" is a package management client for the Paludis
<http://paludis.pioto.org/> package manager available for use with both
the Exherbo Linux <http://exherbo.org/> and Gentoo Linux
<http://gentoo.org/> Distributions.

This module is designed as a syntactic sugar wrapper for that client to
minimize development time and clarify code.

    my $cave = Cave::Wrapper->new();
    my @ids = $cave->print_ids(qw( --matches  dev-lang/perl ));

INSTALLATION

This is a Perl module distribution. It should be installed with whichever
tool you use to manage your installation of Perl, e.g. any of

  cpanm .
  cpan  .
  cpanp -i .

Consult http://www.cpan.org/modules/INSTALL.html for further instruction.
Should you wish to install this module manually, the procedure is

  perl Makefile.PL
  make
  make test
  make install

COPYRIGHT AND LICENSE

This software is copyright (c) 2015 by Kent Fredric
<kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Revision history for Cave-Wrapper

1.000002 2017-03-07T13:19:11Z a9c406f

1.000001 2017-03-07T13:17:37Z a954fdc
 - Bugfix: Avoid test failures with -Ddefault_inc_excludes_dot
 - Internals: stub methods not initialized until after the first instance ( to make ENV setup easier )
 - Bugfix: Don't call ->make_immutable in Moo .. this inflates to Moose.

 [Dependencies::Stats]
 - Dependencies changed since 1.000000, see misc/*.deps* for details
 - develop: +4 ↑2 (suggests: ↑2)
 - test: +2 ↓1 (recommends: +1)

1.000000 2015-06-22T21:55:52Z 0c76760
 [00 Major]
 - Changes in this release may affect hand-operated downstream packagers.

 [01 Version Scheme Change]
 - Version scheme is shortening mantissa to 6 places for consistency.
 - https://questhub.io/realm/perl/quest/5388d3a9bbd0be2224000129
 - This is the primary reason for the 1.x bump

 [Dependencies::Noteworthy]
 - Support level now set for perl 5.6

 [Dependencies::Stats]
 - Dependencies changed since 0.01100100, see misc/*.deps* for details
 - configure: +1 ↓1 (recommends: ↑1)
 - develop: +15 ↑3 -6 (recommends: +1, suggests: +2 -1)
 - runtime: ↓1 -1 (recommends: +1)
 - test: +2 -3 (recommends: ↑2 ↓1 -1)

0.01100100 2014-04-11T04:19:55Z
 [00 Major]
 - Toolkit switched to EUMM
 - cave check done in Makefile.PL
 - implement in Moo and Sub::Install instead of Moose

 [Dependencies::Stats]
 - Dependencies changed since 0.01000004, see misc/*.deps* for details
 - build: -1
 - configure: +2 -1 (recommends: +1)
 - develop: +43 -2 (recommends: -1, suggests: +1 -1)
 - runtime: +3 ↑1 -1
 - test: +4 ↓1 -2 (recommends: +4)

0.01000004 2013-05-10T08:23:28Z
 - Maintenance Release

 [Dependencies::Stats]
 - Dependencies changed since 0.01000003, see misc/*.deps* for details
 - build: ↑1
 - configure: ↑1
 - develop: +6 (recommends: ↑1, suggests: ↑1)
 - runtime: +1
 - test: ↑1

 [Documentation]
 - Update copyright year
 - Reindent license
 - Update license contact address
 - add README.mkdn

 [Meta]
 - bugtracker to github issues.

 [Packaging]
 - use test_requires from MB 0.4004

0.01000003 2011-10-22T13:24:52Z
 - Maintenance Release

 [Dependencies::Stats]
 - Dependencies changed since 0.01000002, see misc/*.deps* for details
 - develop: +1 -1 (recommends: +1 -1, suggests: +1 -1)
 - runtime: +2
 - test: -1

 [Internals]
 - add $AUTHORITY to packages
 - $VERSION no longer set during BUILD
 - Minor style change with FileHandle handling.

 [Packaging]
 - Ship .perltidyrc
 - Move Changelog form to CPAN::Changes.
 - Update LICENSE file ( Year, Indentation, Address )
 - Move author/release tests to xt/
 - GIT uris now https://
 - Move to Git versioning
 - Move to @Author::KENTNL
 - add x_authority

 [Tests]
 - 00-compile.t regenerated.

0.01000002 2010-12-11T08:14:59Z
 - First version.


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.009.
use strict;
use warnings;

# inserted by Dist::Zilla::Plugin::CheckBin 0.007
use Devel::CheckBin;
check_bin('cave');

use 5.006;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "A Wrapper to the Paludis 'cave' Client.",
  "AUTHOR" => "Kent Fredric <kentnl\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "Devel::CheckBin" => 0,
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Cave-Wrapper",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "Cave::Wrapper",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Moo" => 0,
    "Sub::Install" => 0,
    "namespace::autoclean" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Path::Tiny" => 0,
    "Test::Fatal" => 0,
    "Test::More" => 0,
    "Test::TempDir::Tiny" => 0
  },
  "VERSION" => "1.000002",
  "test" => {
    "TESTS" => "t/*.t t/00-compile/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "Moo" => 0,
  "Path::Tiny" => 0,
  "Sub::Install" => 0,
  "Test::Fatal" => 0,
  "Test::More" => 0,
  "Test::TempDir::Tiny" => 0,
  "namespace::autoclean" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


lib/Cave/Wrapper.pm
use 5.006;    # our
use strict;
use warnings;

package Cave::Wrapper;

# ABSTRACT: A Wrapper to the Paludis 'cave' Client.

our $VERSION = '1.000002';

our $AUTHORITY = 'cpan:KENTNL'; # AUTHORITY
























































































use Moo;
use Sub::Install;
use namespace::autoclean;
use Carp qw();

no Moo;

sub _cave_exec_to_list {
  my @args = @_;
  my (@output);
  {
    ## no critic ( ProhibitPunctuationVars )
    open my $fh, q{-|}, 'cave', @args or Carp::croak("Error executing 'cave': $@ $? $!");
    @output = <$fh>;
    close $fh or Carp::carp("Closing 'cave' returned an error: $@ $? $!");
  }
  chomp for @output;
  return @output;
}

my %collisions = map { $_ => 1 } qw( import );

our $_COMMANDS_INITIALIZED;

sub _ensure_initialized {
  return if $_COMMANDS_INITIALIZED;
  for my $command ( _cave_exec_to_list( 'print-commands', '--all' ) ) {
    my $method = $command;
    ## no critic ( RegularExpressions )
    $method =~ s{-}{_}g;
    if ( exists $collisions{$command} ) {
      $method = 'cave_' . $method;
    }
    ## no critic (Subroutines::ProhibitCallsToUnexportedSubs)
    Sub::Install::install_sub(
      {
        code => sub {
          shift;
          return _cave_exec_to_list( $command, @_ );
        },
        as   => $method,
        into => __PACKAGE__,
      },
    );
  }
  $_COMMANDS_INITIALIZED = 1;
  return;
}





sub BUILD { return _ensure_initialized }

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Cave::Wrapper - A Wrapper to the Paludis 'cave' Client.

=head1 VERSION

version 1.000002

=head1 DESCRIPTION

C<cave> is a package management client for the L<Paludis|http://paludis.pioto.org/> package manager available for use with both
the L<Exherbo Linux|http://exherbo.org/> and L<Gentoo Linux|http://gentoo.org/> Distributions.

This module is designed as a syntactic sugar wrapper for that client to minimize development time and clarify code.

    my $cave = Cave::Wrapper->new();
    my @ids = $cave->print_ids(qw( --matches  dev-lang/perl ));

=head1 METHODS

Methods are generated entirely at run-time by introspecting the output from C<cave print-commands --all> and then
generating the appropriate methods. This is mostly because we don't want to have to cut a new release every time
paludis produce a new release I<just> to avoid breaking code.

=head1 CAVEATS

=head2 Naming Collisions

There exists 1 command we cannot perform a native mapping for, and its due to a perlism, and that is C<import>.

For now, this is named C<cave_import> instead.

=head2 Hyphenated Commands

Hyphenated commands can't be used as method names in Perl, so we've translated the hyphens to underscores
in the method names.

i.e.: if you wanted C<print-ids> you now want C<print_ids>

=head2 Slightly Under-powered

This is a first-pass "Just get it working" implementation at this time, and is reasonably useful for the print_ family of commands
the cave client provides. However, you probably do not wish to use it for more complex things like calling C<cave resolve> as it
might cause you untold sorrows while it silently buffers into a growing array and then spews its contents when its finished.

=head1 TODO

One day we'd like to have a sweeter syntax, like

    $cave->print_ids({ matches => 'dev-lang/perl' })

or

    $cave->print_ids({ matches => [ 'dev-lang/perl' , 'dev-lang/python' ]});

However, there are a few problems and questions to be answered, which are not a problem with the existing
syntax but would be a problem with a possible alternative syntax.

=over 4

=item * Toggle Switches

There are a lot of toggle switches that don't take a parameter, and while we could just do

    $cave->print_commands({ all => 1 });

That means we have to get rid of the '1' before we pass the command to C<cave>, and that is going to be difficult to
do without needing tight coupling. Not to mention how to handle C<< all => 2 >> and C<< all => 1 >>.

=item * Fixed Order operations.

Some C<cave> functions require operators to be ordered, so if you needed to do this

    cave foobar --matching foo --not --matching bar

having a structure

    $cave->foobar({ matching => [ 'foo', 'bar' ]} , '--not' )

would obviously not work.

    $cave->foobar([ matching => 'foo', not => matching => 'bar' ])

or anything not using a hash is going to be equally confusing, especially as we can now no longer tell what
is a key and what is a value, so adding '--' to the front of them becomes impossible.

=back

=for Pod::Coverage BUILD

=head1 AUTHOR

Kent Fredric <kentnl@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2017 by Kent Fredric <kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


maint/perlcritic.rc.gen.pl
#!/usr/bin/env perl
## no critic (Modules::RequireVersionVar)

# ABSTRACT: Write an INI file from a bundle

use 5.008;    # utf8
use strict;
use warnings;
use utf8;

our $VERSION = 0.001;

use Carp qw( croak carp );
use Perl::Critic::ProfileCompiler::Util qw( create_bundle );

## no critic (ErrorHandling::RequireUseOfExceptions)
my $bundle = create_bundle('Example::Author::KENTNL');
$bundle->configure;

my @stopwords = (qw(
  Paludis Gentoo Exherbo paludis perlism TODO
));
for my $wordlist (@stopwords) {
  $bundle->add_or_append_policy_field( 'Documentation::PodSpelling' => ( 'stop_words' => $wordlist ) );
}

#$bundle->add_or_append_policy_field(
#  'Subroutines::ProhibitCallsToUndeclaredSubs' => ( 'exempt_subs' => 'String::Formatter::str_rf' ), );

#$bundle->remove_policy('ErrorHandling::RequireCarping');
#$bundle->remove_policy('NamingConventions::Capitalization');

my $inf = $bundle->actionlist->get_inflated;

my $config = $inf->apply_config;

{
  open my $rcfile, '>', './perlcritic.rc' or croak 'Cant open perlcritic.rc';
  $rcfile->print( $config->as_ini, "\n" );
  close $rcfile or croak 'Something fubared closing perlcritic.rc';
}
my $deps = $inf->own_deps;
{
  open my $depsfile, '>', './perlcritic.deps' or croak 'Cant open perlcritic.deps';
  for my $key ( sort keys %{$deps} ) {
    $depsfile->printf( "%s~%s\n", $key, $deps->{$key} );
    *STDERR->printf( "%s => %s\n", $key, $deps->{$key} );
  }
  close $depsfile or carp 'Something fubared closing perlcritic.deps';
}



Build.PL
use Devel::CheckBin;
check_bin('cave');


# This file was automatically generated by Dist::Zilla::Plugin::ModuleBuild v5.015.
use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.4205"
  },
  "configure_requires" => {
    "Devel::CheckBin" => 0,
    "Module::Build" => "0.4205"
  },
  "dist_abstract" => "A Wrapper to the Paludis 'cave' Client.",
  "dist_author" => [
    "Kent Fredric <kentnl\@cpan.org>"
  ],
  "dist_name" => "Cave-Wrapper",
  "dist_version" => "0.01000005",
  "license" => "perl",
  "module_name" => "Cave::Wrapper",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Carp" => 0,
    "Moose" => 0,
    "namespace::autoclean" => 0,
    "perl" => "5.006",
    "strict" => 0,
    "warnings" => 0
  },
  "script_files" => [],
  "test_requires" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec::Functions" => 0,
    "List::Util" => 0,
    "Test::Fatal" => 0,
    "Test::More" => "1.001003",
    "version" => 0
  }
);


my %fallback_build_requires = (
  "ExtUtils::MakeMaker" => 0,
  "File::Spec::Functions" => 0,
  "List::Util" => 0,
  "Module::Build" => "0.4205",
  "Test::Fatal" => 0,
  "Test::More" => "1.001003",
  "version" => 0
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;
